### PR TITLE
For #9189 - Pocket articles support

### PR DESCRIPTION
**Pull request by Mugurell**
Since the previous version of this component was not in use anymore it was used
as a starting point for the newly needed functionality - easily offering Pocket
recommended articles to interested clients.

The new version offers a simple API for downloading, caching and refreshing in
the background the list of Pocket articles that is offered to clients.

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Comment by pocmo** [comment]
Random drive-by question. I saw you created `feature-pocket-recommendations`. There's also `service-pocket` as a generic pocket component, derived from the Amazon apps afaik. Is it possible to merge that together?

**Comment by Mugurell** [comment]
> Random drive-by question. I saw you created `feature-pocket-recommendations`. There's also `service-pocket` as a generic pocket component, derived from the Amazon apps afaik. Is it possible to merge that together?

Not sure about the splitting and what each layer of AC should be, how should they all be stacked. I'm ok with any recommendations.
I saw `service-pocket` (based on just the naming also) as offering an abstraction over downloading (and returning) a list of items from Pocket, a general piece of functionality that could have many use cases.
Which is then used by the new `feature-pocket-recommendations` which serves a particular scope and is much more specific, imposing a certain way of getting the data from Pocket.

Having both as a single module would make things easier (for my patch) but I also see reasons for keeping these separate.
I see @mcomella  being the owner of `service-pocket`, maybe he knows more about the plans for this component, if it would be ok to be absorbed in a new one..?

**Comment by mcomella** [comment]
(repasting something I sent to Petru b/c it's useful context here) I don’t believe any of the code is in use in `service-pocket`. They disabled the Pocket video recommendation endpoint and the other code was used for a prototype and external folks don’t have the API keys to use it. I’m fine with anything – this getting merged into your new component, your new component getting merged here, or even with this getting deprecated – we’re not actively using the code and I see no near future in which we do use it so I don’t think it’s worth effort to make sure we keep it working.

I see I’m the code owner in the service-pocket module – I don’t actually have cycles to review this. I think it’s okay for someone else to review this – I assume it’s just a normal get-and-serialize-json-data PR. The general advice I’d mention is:
- Make sure the retry code is correct, if it exists. We once had a bug where we retried many times per second, causing denial-of-service behavior and that’s really bad because we can’t guarantee users update to the latest version
- From a performance perspective, if the UI doesn’t require on-demand updates, it’s simpler and probably more effective to do this via WorkManager so that users won’t be waiting on fetches and fetches won’t slow down other requests in the app

**Comment by pocmo** [comment]
Makes sense, thanks Mike!

I think in any case we should still end up with only one component providing Pocket functionality. If we need more than it is very likely to share code. Maybe you could turn this into the new `service-pocket` ("service" because it primarily provides access to a backend service) and we get rid of the existing code, since - as mentioned - it's not in use and unlikely to get used unfortunately.

**Comment by Mugurell** [comment]
Yeah, thinking more about it, maybe it makes more sense for this to be just a service.
I started this new `feature-pocket..` as working similar to `feature-top-sites`.
All the business logic being in the ac, clients interested in using the "pocket feature" having to then implement the "Pocket View" interface that would ensure them adhering to a set of specific actions.
But if everything this that this component would do is just allow querying a list of pocket articles a service might make more sense.
And if in the future more functionality is to be added it can then be easily refactored in a new feature.

**Comment by Mugurell** [comment]
Sorry this got so big and is a single commit, but should be straightforward and easy to follow.

I'd say the most interesting part is exactly the entry point to this functionality - PocketStoriesService#startPeriodicStoriesRefresh
Because transparent data refresh in the background is important and a PeriodicWorkRequest was used for this we also need to know when to stop this work.
As such what I recommended is to use PocketStoriesService#startPeriodicStoriesRefresh and PocketStoriesService#stopPeriodicStoriesRefresh as higher up in the client app as possible (preferable in the Application or the single Activity).
Open to other suggestions regarding this but testing it lightly on Fenix it seems to work ok.

File: components/service/pocket/src/main/java/mozilla/components/service/pocket/api/PocketEndpoint.kt
Comment (jonalmeida) [inline components/service/pocket/src/main/java/mozilla/components/service/pocket/api/PocketEndpoint.kt:55]: Probably best to remove this default and the `lib-fetch-httpurlconnection` in the `build.gradle`.

File: components/service/pocket/src/test/java/mozilla/components/service/pocket/api/PocketEndpointRawTest.kt
Comment (csadilek) [inline components/service/pocket/src/test/java/mozilla/components/service/pocket/api/PocketEndpointRawTest.kt:48]: nit: let's import `Mockito.when`

File: components/service/pocket/src/test/java/mozilla/components/service/pocket/api/PocketEndpointRawTest.kt
Comment (csadilek) [inline components/service/pocket/src/test/java/mozilla/components/service/pocket/api/PocketEndpointRawTest.kt:70]: nit: import `Assert.`

File: components/service/pocket/src/test/java/mozilla/components/service/pocket/api/PocketEndpointTest.kt
Comment (csadilek) [inline components/service/pocket/src/test/java/mozilla/components/service/pocket/api/PocketEndpointTest.kt:49]: nit: import `Mockito.when`. You can also use `whenever` to avoid the back ticks.

File: components/service/pocket/src/main/java/mozilla/components/service/pocket/stories/update/RefreshPocketWorker.kt
Comment (csadilek) [inline components/service/pocket/src/main/java/mozilla/components/service/pocket/stories/update/RefreshPocketWorker.kt:41]: @Amejia481 and I ran into this before and found that returning `failure` here actually prevents the periodic task from running again, which is not what you want in this case. We'd have to verify again, but then you'd want to return `retry` or `success` here instead.

**Review by csadilek — CHANGES_REQUESTED** [review]
Looks great, @Mugurell. I found a few things to fix and some nits to address below.

Two additional topics to discuss, not covered below:
- I would prefer if we didn't mix `Stories` and `Articles` as much. It was quite confusing to read, as it's all the same, but we're switching back and forth quite a bit in the API. Should we just use `Story` as a concept and skip `Article`. If not, what's the difference?
- I am not convinced we should use Room (or a sqlite database at all) to cache pocket stories. It's quite some extra complexity and we're really only querying all data at once i.e. we're not using any of the query functionality provided by the DB. It seems just storing a flat file of the last response would be much easier and efficient? Do you foresee more complicated use cases where we would need structured queries?

File: components/service/pocket/src/main/java/mozilla/components/service/pocket/stories/update/RefreshPocketWorker.kt
Comment (Amejia481) [reply components/service/pocket/src/main/java/mozilla/components/service/pocket/stories/update/RefreshPocketWorker.kt:41]: Yeah,  you can find some docs [here](https://developer.android.com/reference/androidx/work/ListenableWorker.Result#failure()), additionally we did something for `AddonUpdaterWorker` where [we stopped a periodic work](https://github.com/mozilla-mobile/android-components/blob/a9e5357a4ea19ceb07de96b3ccf5e72a56d3b427/components/feature/addons/src/main/java/mozilla/components/feature/addons/update/AddonUpdater.kt#L591) from getting executed when add-on was not installed anymore.

File: docs/changelog.md
Comment (rocketsroger) [inline docs/changelog.md:24]: This will have to be moved to the current A-C development version.

File: components/service/pocket/src/test/java/mozilla/components/service/pocket/stories/PocketRecommendationsRepositoryTest.kt
Comment (rocketsroger) [inline components/service/pocket/src/test/java/mozilla/components/service/pocket/stories/PocketRecommendationsRepositoryTest.kt:60]: Please add blank line after this.  Same as the other files.  Thanks,

File: components/service/pocket/src/test/java/mozilla/components/service/pocket/helpers/Params.kt
Comment (rocketsroger) [inline components/service/pocket/src/test/java/mozilla/components/service/pocket/helpers/Params.kt:12]: Add extra line.  Thanks

File: components/service/pocket/build.gradle
Comment (Mugurell) [reply components/service/pocket/build.gradle:43]: I saw Jonathan recommending the same.
I would be curious as to why.
I wanted this component to be plug and play, as easy to use as possible and so for me it makes more sense for it to have everything it needs, have full control over it's dependencies and internal behavior while potentially allowing clients to override part of.
As a client (let's say other than Fenix) only interested in a list of Pocket articles I'd prefer to not care about how the download is made in the background and I may not even have an implementation of `concept.fetch.Client` to pass to this component.

File: components/service/pocket/README.md
Comment (Mugurell) [reply components/service/pocket/README.md:8]: Thank you!

File: components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketRecommendedArticle.kt
Comment (Mugurell) [reply components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketRecommendedArticle.kt:19]: Thank you!

File: components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketStoriesService.kt
Comment (Mugurell) [reply components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketStoriesService.kt:17]: Thank you, will use `Story` everywhere.

File: components/service/pocket/build.gradle
Comment (jonalmeida) [reply components/service/pocket/build.gradle:43]: To add to the above, for non-Fenix applications as well, making a conscience decision to use their own provided HTTP client will make it clear what control they have to maintain; avoids creating multiple clients unknowingly.

We provide non-GV fetch clients for non-Mozilla consumers, but _how_ that resource is instantiated and maintained _needs_ to be decided by them. This avoids creating a new client for every service/ui/feature component unintentionally as well.

**Review by csadilek — COMMENTED** [review]
> @csadilek Regarding serializing the stories in a File versus a db I'd prefer the db route just to get the ACID guarantees ....  synchronize read/writes in the File. We don't use any advanced queries or relations but the code is pretty clear, straight-forward and saves us from a bit of headaches.

@Mugurell Right, but we also have `AtomicFile` and ext. functions to provide that, plus we're already using that in similar instances where we fetch and cache from a server e.g. in our `AddonCollectionProvider`.

More generally, I think we should only take the overhead of a full DB if we intend to use its powers, which we may or may not want here in the future. The fact that we are currently throwing all records away and inserting/querying them all together is a smell. We're basically using the DB as a flat file here.

I will leave it up to you :)

File: components/service/pocket/src/main/java/mozilla/components/service/pocket/api/PocketGlobalArticleRecommendation.kt
Comment (Mugurell) [reply components/service/pocket/src/main/java/mozilla/components/service/pocket/api/PocketGlobalArticleRecommendation.kt:25]: Maybe having just `A Pocket recommended story.` as description for the type offered to clients is enough.

File: components/service/pocket/src/test/java/mozilla/components/service/pocket/api/PocketEndpointRawTest.kt
Comment (Mugurell) [reply components/service/pocket/src/test/java/mozilla/components/service/pocket/api/PocketEndpointRawTest.kt:48]: Totally!

File: components/service/pocket/src/main/java/mozilla/components/service/pocket/stories/update/RefreshPocketWorker.kt
Comment (Mugurell) [reply components/service/pocket/src/main/java/mozilla/components/service/pocket/stories/update/RefreshPocketWorker.kt:41]: Thank you!

**Comment by Mugurell** [comment]
> More generally, I think we should only take the overhead of a full DB if we intend to use its powers, which we may or may not want here in the future. The fact that we are currently throwing all records away and inserting/querying them all together is a smell. We're basically using the DB as a flat file here.

Thank you!
Will look into switching to using an AtomicFile.

File: components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketStoriesService.kt
Comment (Mugurell) [reply components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketStoriesService.kt:27]: Guess it's a matter of perspective 😅.
Let's consider the Android Application at the top and then will use "as high as possible".

File: components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketRecommendedArticle.kt
Comment (csadilek) [reply components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketRecommendedArticle.kt:17]: Yeah let's remove if always empty now.

File: components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketRecommendedArticle.kt
Comment (csadilek) [reply components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketRecommendedArticle.kt:16]: 👍

**Comment by Mugurell** [comment]
I've updated the PR based on your feedback, thank you all!
Did some small tests in Fenix to see that the entire flow works, all good.

Updates just pushed:
- Updated various naming issues all throughout the patch.
- Switched to using an AtomicFile to persist the stories locally, all seems to work okay and the code is simpler.
- Removed `HttpURLConnectionClient` default from code (implementation still used in tests though). Initially all configuration was injected in the Worker but there is no way to pass the Client as `extra` data to the Worker so I went with caching both the Pocket api key and the Client in `PocketStoriesUseCases`'s companion object so that when the Worker uses an usecase to download new data we'll have the passed in Client available.
- Updated / added tests accordingly.

**Review by csadilek — APPROVED** [review]
Thanks, @Mugurell. I found a few things to fix up before landing, but ✅ 🚢 .

File: components/service/pocket/src/test/java/mozilla/components/service/pocket/api/PocketJSONParserTest.kt
Comment (Mugurell) [reply components/service/pocket/src/test/java/mozilla/components/service/pocket/api/PocketJSONParserTest.kt:21]: This was used in the initial version to convert the Pocket JSON to out Kotlin model classes.
If we're persisting the JSON this is not needed anymore but I forgot to delete the file.
Thank you for spotting it!

File: components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketArticlesConfig.kt
Comment (Mugurell) [reply components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketArticlesConfig.kt:26]: Thank you!
Renamed it now.

File: components/service/pocket/build.gradle
Comment (Mugurell) [reply components/service/pocket/build.gradle:18]: Indeed. Thank you!

**Review by csadilek — APPROVED** [review]
🚢

**Pull request merged**
